src/lxc/conf.c

```diff
@@ -757,79 +788,80 @@ static int lxc_mount_auto_mounts(struct lxc_conf *conf, int flags, struct lxc_ha
 	}
 
 	if (flags & LXC_AUTO_SHMOUNTS_MASK) {
-		char *new_mount;
-		size_t len_mount;
+		size_t len_cont;
+		char *full_cont_path;
+		struct stat sb;
 		int ret = -1;
 
+		/* Construct the shmount path under the container root */
 		/* +1 for slash */
-		size_t len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
+		len_cont = strlen(conf->rootfs.mount) + 1 + strlen(conf->lxc_shmount.path_cont);
 		/* +1 for the terminating '\0' */
-		char *full_cont_path = malloc(len_cont + 1);
+		full_cont_path = malloc(len_cont + 1);
 		if(!full_cont_path) {
 			SYSERROR("Not enough memory");
 			return -ENOMEM;
 		}
-
 		ret = snprintf(full_cont_path, len_cont + 1, "%s/%s", conf->rootfs.mount, conf->lxc_shmount.path_cont);
 		if (ret < 0 || ret >= len_cont + 1) {
 			SYSERROR("Failed to create filename");
-			goto out_free_path;
+			free(full_cont_path);
+			return -1;
 		}
 
+		/* Check if shmount point is already set up */
 		if(is_shared_mountpoint(conf->lxc_shmount.path_host)
-				&& is_shared_mountpoint(full_cont_path)) {
-			/* Already set up */
-			ret = 0;
-			goto out_free_path;
+				&& !stat(full_cont_path, &sb) && S_ISDIR(sb.st_mode)) {
```

Comment:
Hm, don't you need to call `stat()` on `path_host` and not on `full_cont_path`? How about:
```
if (is_shared_mountpoint(conf->lxc_shmount.path_host) {
        int ret;

        ret = stat(conf->lxc_shmount.path_host, &sb);
        if (ret == 0 && S_ISDIR(sb.st_mode)) {
                INFO("Path \"%s\" is already MS_SHARED. Reusing", conf->lxc_shmount.path_host);
                goto finalize;
        }
}
```

Reply:
Oh, right, I see. But do we need this check at all then? `mount` seems to return `ENOTDIR` if `path_host` was not a directory

Reply:
Sure, if that simplifies things. :)